grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,185 @@
+      # that can be listed on the dropdown-menu called "Metric".
+      response = AVAILABLE_METRICS
+    else:
+      # Grafana request issued on Variables > New/Edit page for variables of type query.
+      # Grafana expectes the list of possible values of the variable.
+      # At the moment, GRRafana doesn't support such variables, so it returns no possible values.
+      response = []
+    return JSONResponse(response=json.dumps(response), mimetype=JSON_MIME_TYPE)
+
+  def OnQuery(self, request: JSONRequest) -> JSONResponse:
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    json_data = request.json
+    requested_client_id = _ExtractClientIdFromVariable(
+        json_data)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in json_data["targets"]]
+    response = _FetchDatapointsForTargets(requested_client_id,
+                                          json_data["maxDataPoints"],
+                                          requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype=JSON_MIME_TYPE)
+
+  def OnAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+Datapoint = Tuple[float, int]
+Datapoints = List[Datapoint]
+TargetWithDatapoints = Dict[Text, Datapoints]
+
+
+def _FetchDatapointsForTargets(
+    client_id: Text, limit: int,
+    targets: List[Text]) -> List[TargetWithDatapoints]:
+  """Fetches a list of <datapoint, timestamp> tuples for each target metric from Fleetspeak database."""
+  records_list = fleetspeak_utils.FetchClientResourceUsageRecordsFromFleetspeak(
+      client_id, limit)
+  response = []
+  for target in targets:
+    datapoints_for_single_target = _CreateDatapointsForTarget(
+        target, records_list)
+    target_datapoints_dict = cast(TargetWithDatapoints, {
+        "target": target,
+        "datapoints": datapoints_for_single_target
+    })
+    response.append(target_datapoints_dict)
+  return response
+
+
+def _CreateDatapointsForTarget(
+    target: Text, records_list: List[resource_pb2.ClientResourceUsageRecord]) -> Datapoints:
+  if target == "mean_user_cpu_rate":
+    record_values = [record.mean_user_cpu_rate for record in records_list]
+  elif target == "max_user_cpu_rate":
+    record_values = [record.max_user_cpu_rate for record in records_list]
+  elif target == "mean_system_cpu_rate":
+    record_values = [record.mean_system_cpu_rate for record in records_list]
+  elif target == "max_system_cpu_rate":
+    record_values = [record.max_system_cpu_rate for record in records_list]
+  elif target == "mean_resident_memory_mib":
+    record_values = [record.mean_resident_memory_mib for record in records_list]
+  elif target == "max_resident_memory_mib":
+    record_values = [record.max_resident_memory_mib for record in records_list]
+  return [(v, r.server_timestamp.seconds * 1000)
```

Comment:
Your function returns milliseconds, but uses only second-precision. Please extract the millisecond part from `server_timestamp.nanos`.